### DISCUSSION TITLE
Would this be possible to integrate with Klein (twisted-based "Flask" alternative)

### DISCUSSION THREAD

**Discussion by Column01** [LiBa001/discord-interactions-wrapper#20] [General]
Would it be possible to integrate this with [Klein](https://klein.readthedocs.io/en/latest/) instead of Flask using the existing flask extension code? Klein uses the same routing system that Flask does and provides twisted support (if you aren't sure what that is, it's sort of like `asyncio` but an entirely separate framework).

I have an existing application built on twisted and would like to integrate a discord plugin into it. It requires the ability to run inside twisted, and I'm hoping that this can be done with Klein OOTB with your existing code. Looking at the integration, it doesn't seem possible. An Extension would likely need to be added as far as I can tell.

**Comment by Column01** [comment]
I've started work on making a heavily inspired wrapper that will allow for twisted web servers (not Klein directly). Will 1000% link to your project in the readme once it goes live (AKA once it's functional and tested)

**Comment by Column01** [comment]
https://github.com/Column01/twisted-discord-interactions It's public now, but still WIP. Sorry to clutter your repo with a discussion about this, but I figured I'd ask in case you wanted to add support.

**Reply by LiBa001** [reply]
Thanks! I don't know anything about Klein and twisted, but I'll look into your code once I got more time. I'm currently rather busy, sorry.

**Reply by Column01** [reply]
No worries! I plan to flesh it out more and maybe just work as a standalone library. The twisted ecosystem is rather confusing if you've never worked with it and it has poor documentation.

Twisted is essentially asyncio before it was cool (except it uses threads). It has a "reactor" instead of an event loop which lets you spawn processes and uses various networking structures to allow for deferrals (aka `await`) for asynchronous programming. I don't necessarily think it fits with your current project.

**Comment by LiBa001** [comment]
@Column01 btw, did you look at the dev branch of this project? Because I was actually designing a more general extension approach with a `BaseExtension` abstract base class which can be inherited from and implemented to use a specific framework. The extension is then basically only responsible for using the framework-specific way of registering a route and retrieving all the request data, using it to call the predefined `_handle_interaction` method and convert/send back the returned interaction response.

The plan was to maybe also have a FastAPI extension. So we'd have `ext.flask`, `ext.fastapi` and possibly `ext.klein` or something.
As I said, I don't really know Klein and twisted, but from what I've seen, I don't really know why it wouldn't work.
If you're still interested, you could try that and maybe make a pull request later on. Shouldn't take too long, since all the interaction logic is already handled by the base extension.

**Reply by Column01** [reply]
Well part of the problem is twisted's ecosystem uses deferrals and I'm not sure how it could be implemented without having dedicated code for twisted, either way it needs code added to work and the average person wouldn't be capable afaik. My project is on hold atm for a variety of reasons so I haven't finished my library. I will look at the dev branch, that is probably a better option since a lot of the code already exists from you.
